### DISCUSSION TITLE
Custom ScaleKernel with ARD prior (batched GPs)

### DISCUSSION THREAD

**Discussion by coding-ecoli** [cornellius-gp/gpytorch#2537] [Q&A] [answered]
Hi,

Thank you for making GPytorch!! I really enjoy working with it! 

I have a question regarding custom decorative kernels: I have read this tutorial: https://docs.gpytorch.ai/en/stable/examples/00_Basic_Usage/Implementing_a_custom_Kernel.html but I am still not so sure how to apply these on decorative kernels. What is the best practice for writing a custom decorative Kernel, e.g. a ```ScaleKernel()```? Can we just inherit from the ```ScaleKernel``` class and rewrite the ```forward()```? In the base ```ScaleKernel()``` class it is already possible to define a prior over the output scale but I want to play a little with an ARD prior over the output scales, $K_{scaled} = \theta *K_{base}$, where:

$\alpha \sim Gamma(1.0, 1.0)$
$g(x) :=$ some constraint, e.g. non-negative
$\theta \sim g(N(0, \frac{1}{\alpha}))$

One simpler custom ```ScaleKernel``` example would be $K_{scaled} = \frac{1}{\theta} *K_{base}$:

I have inherited from ScaleKernel
```
class CustomScaleKernel(ScaleKernel):
             ...
             def forward(...):
                    outputscale = 1. /self.outputscale # I just added this 1./ 
```

Is that the correct way?

Thanks and warm regards,
Wangjun

**Comment by jacobrgardner** [comment, marked as answer]
Unless you specifically want to be working with an outputscale I'm just a different way, I would honestly just copy ScaleKernel and implement what you need it to do rather than extending the original class. The only real reason to extend ScaleKernel would be code reuse -- is there code in ScaleKernel that you'd still be using?

For your very simple example, I'd just extend ScaleKernel and override the outputscale getters and setters to just return 1/ the outputscale:
https://github.com/cornellius-gp/gpytorch/blob/9551eba889adf835b69cfd86e9a5d584fb61cdcc/gpytorch/kernels/scale_kernel.py#L96

**Reply by coding-ecoli** [reply]
Thank you for your answer!
